content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -1,38 +1,408 @@
+but in Ada we go to a package named `System` to get this information.
 
-    Complete section!
+Clearly, these implementation values will vary with the hardware, so the
+package declares constants with implementation-defined values. The names
+of the constants are what's portable, you can count on them being the
+same in any Ada implementation.
+
+However, vendors can add implementation-defined declarations to the
+language-defined content in package System. You might require some of
+those additions, but portability could then suffer when moving to a new
+vendor's compiler. Try not to use them unless it is unavoidable. Ideally
+these additions will appear in the private part of the package, so the
+implementation can use them but application code cannot.
+
+For examples of the useful, language-defined constants, here are those
+for the numeric limits of an Ada compiler for an Arm 32-bit SoC:
+
+.. code-block:: ada
+
+   Min_Int               : constant := Long_Long_Integer'First;
+   Max_Int               : constant := Long_Long_Integer'Last;
+
+   Max_Binary_Modulus    : constant := 2 ** Long_Long_Integer'Size;
+   Max_Nonbinary_Modulus : constant := 2 ** Integer'Size - 1;
+
+   Max_Base_Digits       : constant := Long_Long_Float'Digits;
+   Max_Digits            : constant := Long_Long_Float'Digits;
+
+   Max_Mantissa          : constant := 63;
+   Fine_Delta            : constant := 2.0 ** (-Max_Mantissa);
+
+`Min_Int` and `Max_Int` supply the most-negative and most-positive
+integer values supported by the machine.
+
+`Max_Binary_Modulus` is the largest power allowed as the modulus of a
+modular type definition.
+
+But a modular type need not be defined in terms of powers of two. An
+arbitrary modulus is allowed, as long is it is not bigger than the
+machine can handle. That's specified by `Max_Nonbinary_Modulus`, the
+largest (positive) value allowed as the modulus of a modular type
+definition.
+
+`Max_Base_Digits` is the largest value allowed for the requested decimal
+precision in a floating-point type's definition.
+
+We won't go over all of the above, you get the idea. Let's examine the
+more important contents.
+
+Two of the most frequently referenced constants in System are the
+following, especially the first. (The values here are again for the Arm
+32-bit SoC):
+
+.. code-block:: ada
+
+   Storage_Unit : constant := 8;
+   Word_Size    : constant := 32;
+
+`Storage_Unit` is the number of bits per memory storage element. Storage
+elements are the components of memory cells, and typically correspond to
+the individually addressable memory elements. A "byte" would correspond
+to a storage element with the above constant value.
+
+Consider a typical idiom for determining the number of whole storage
+elements an object named X occupies:
+
+.. code-block:: ada
+
+  Units : constant Integer := (X'Size + Storage_Unit - 1) / Storage_Unit;
+
+Remember that `'Size` returns a value in terms of bits. There are more
```

Comment:
```suggestion
Remember that :ada:`'Size` returns a value in terms of bits. There are more
```

Also, as far as I can see, the concept of `'Size` hasn't been introduced in previous paragraphs yet, so you might want to explain it here.